### PR TITLE
Don't use Uri class for GetRelativePath

### PR DESCRIPTION
**Pull request by BrennanConroy**

### CODE REVIEW

File: test/Microsoft.Framework.Runtime.Tests/PathUtilityFacts.cs [hunk 1]
```diff
@@ -39,5 +39,21 @@ public void IsChildOfDirectoryWorksOnBaseDirWithoutTrailingPathSeparator()
             Assert.True(PathUtility.IsChildOfDirectory(baseDir, childPath));
             Assert.False(PathUtility.IsChildOfDirectory(baseDir, nonChildPath));
         }
+
+        [Fact]
+        public void GetRelativePathReturnsCorrectRelativePaths()
```
Comment (davidfowl) [inline test/Microsoft.Framework.Runtime.Tests/PathUtilityFacts.cs]: Use theories instead and add some tests that handle windows style paths.

**Comment by BrennanConroy** [comment]
Fixed, also the CI's passed just before I changed it

File: src/Microsoft.Framework.Runtime/NuGet/Utility/PathUtility.cs [hunk 2]
```diff
@@ -72,7 +72,58 @@ public static string GetRelativePath(string path1, string path2)
...
+            var path2Segments = path2.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+            // if path1 does not end with / it is assumed the end is not a directory
+            // we will assume that is isn't a directory by ignoring the last split
+            var len1 = path1Segments.Length - 1;
+            var len2 = path2Segments.Length;
+
+            int min = Math.Min(len1, len2);
+            while (min > index)
+            {
+                if (string.Compare(path1Segments[index], path2Segments[index], StringComparison.OrdinalIgnoreCase) != 0)
```
Comment (moozzyk) [inline src/Microsoft.Framework.Runtime/NuGet/Utility/PathUtility.cs]: `StringComparison.OrdinalIgnoreCase` - paths are case sensitive on non-windows

**Comment by moozzyk** [comment]
Honestly, I think we should not have this code. If the Uri class is broken then it may be good as a stopgap but ultimately this should be fixed in the Uri class... Maybe find someone who own Uri and have them confirm this is a bug and maybe they will give you a workaround?

**Comment by BrennanConroy** [comment]
@davidfowl 
Should we get this in?
I filed an issue on CoreFX
https://github.com/dotnet/corefx/issues/1745

**Comment by davidfowl** [comment]
@BrennanConroy no. I'm on the same page as @moozzyk and I'm also worried about regressions.

**Comment by BrennanConroy** [comment]
@davidfowl @moozzyk 
Please review before Monday

File: test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs [hunk 3]
```diff
@@ -40,5 +41,54 @@ public void IsChildOfDirectoryWorksOnBaseDirWithoutTrailingPathSeparator()
             Assert.True(PathUtility.IsChildOfDirectory(baseDir, childPath));
             Assert.False(PathUtility.IsChildOfDirectory(baseDir, nonChildPath));
         }
+
+        public static IEnumerable<object[]> FolderPaths
+        {
+            get {
+                var list =  new List<object[]>
```
Comment (davidfowl) [inline test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs]: You  need some tests with the slashes in the opposite direction.

Comment (davidfowl) [inline test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs]: Also add file paths.

**Comment by BrennanConroy** [comment]
Updated

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs [hunk 4]
```diff
@@ -72,7 +73,72 @@ public static string GetRelativePath(string path1, string path2)
         /// </summary>
         public static string GetRelativePath(string path1, string path2, char separator)
         {
-            return UriUtility.GetPath(GetRelativeUri(path1, path2), separator);
+            if (path1 == null)
```
Comment (moozzyk) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: does this function work if one of the paths (or both) is `string.Empty`?

File: test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs [hunk 5]
```diff
@@ -40,5 +41,56 @@ public void IsChildOfDirectoryWorksOnBaseDirWithoutTrailingPathSeparator()
             Assert.True(PathUtility.IsChildOfDirectory(baseDir, childPath));
             Assert.False(PathUtility.IsChildOfDirectory(baseDir, nonChildPath));
         }
+
+        public static IEnumerable<object[]> FolderPaths
+        {
+            get {
```
Comment (moozzyk) [inline test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs]: nit: put `{` on the next line

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs (hunk 4 above)
Comment (moozzyk) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: also all the tests are using absolute paths - what if a relative path is passed?

File: test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs (hunk 3 above)
Comment (davidfowl) [inline test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs]: Add some tests that try to resolve 2 absolute paths

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs (hunk 4 above)
Comment (BrennanConroy) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: The function needs absolute paths, what does it even mean to get a relative path from two relative paths? Its not possible without assuming they have the same base path.
If string.Empty is passed it should probably error too?

File: test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs (hunk 3 above)
Comment (BrennanConroy) [inline test/Microsoft.Dnx.Runtime.Tests/PathUtilityFacts.cs]: Isn't that what all the tests are doing?

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs (hunk 4 above)
Comment (moozzyk) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: This was my point. You should throw in either case.

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs [hunk 6]
```diff
@@ -145,6 +211,10 @@ public static string GetPathWithBackSlashes(string path)
 
         public static string GetPathWithDirectorySeparator(string path)
         {
+            if (!RuntimeEnvironmentHelper.IsWindows)
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: This is incorrect. This helpers is used in other places to normalize slashes in paths.

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs [hunk 7]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.IO;
+using Microsoft.Dnx.Runtime;
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: Remove this

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs [hunk 8]
```diff
@@ -72,7 +72,72 @@ public static string GetRelativePath(string path1, string path2)
...
+                    break;
+                }
+                ++index;
+            }
+
+            var path = "";
+
+            // check if path2 ends with a non-directory separator and if path1 has the same non-directory at the end
+            if (len1 + 1 == len2 && !string.IsNullOrEmpty(path1Segments[index]) &&
+                string.Compare(path1Segments[index], path2Segments[index], compare) == 0)
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: string.Equals?

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs [hunk 9]
```diff
@@ -72,7 +72,72 @@ public static string GetRelativePath(string path1, string path2)
...
+            var path2Segments = path2.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+            // if path1 does not end with / it is assumed the end is not a directory
+            // we will assume that is isn't a directory by ignoring the last split
+            var len1 = path1Segments.Length - 1;
+            var len2 = path2Segments.Length;
+
+            var min = Math.Min(len1, len2);
+            while (min > index)
+            {
+                if (string.Compare(path1Segments[index], path2Segments[index], compare) != 0)
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: !string.Equals?

**Comment by davidfowl** [comment]
:shipit:

File: src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs [hunk 10]
```diff
@@ -72,7 +72,75 @@ public static string GetRelativePath(string path1, string path2)
         /// </summary>
         public static string GetRelativePath(string path1, string path2, char separator)
         {
-            return UriUtility.GetPath(GetRelativeUri(path1, path2), separator);
+            if (string.IsNullOrEmpty(path1))
+            {
+                throw new ArgumentException(nameof(path1));
```
Comment (moozzyk) [inline src/Microsoft.Dnx.Runtime/NuGet/Utility/PathUtility.cs]: The string you passed is the error message.... The second parameter is the argument name.

**Comment by BrennanConroy** [comment]
046a0595d32357b7ef5c5c58fe877b2225cad8fb

**Pull request closed without merging**